Runtime/AvatarCreator/TemplateFetcher.cs

```diff
@@ -7,38 +7,34 @@ namespace ReadyPlayerMe.AvatarCreator
 {
     public class TemplateFetcher
     {
-        private readonly CancellationToken ctx;
         private readonly AvatarAPIRequests avatarAPIRequests;
-        private readonly List<TemplateData> templates;
 
         public TemplateFetcher(CancellationToken ctx = default)
         {
-            this.ctx = ctx;
             avatarAPIRequests = new AvatarAPIRequests(ctx);
-            templates = new List<TemplateData>();
         }
 
         public async Task<List<TemplateData>> GetTemplates()
         {
-            var avatarTemplates = await avatarAPIRequests.GetTemplates();
-            await GetAllTemplateRenders(avatarTemplates);
-            return templates;
+            return await avatarAPIRequests.GetTemplates();
         }
 
-        private async Task GetAllTemplateRenders(IEnumerable<TemplateData> templateAvatars)
+        public async Task<List<TemplateData>> GetTemplatesWithRenders()
         {
-            var downloadRenderTasks = templateAvatars.Select(GetAvatarRender).ToList();
+            var templates = await avatarAPIRequests.GetTemplates();
+            await FetchTemplateRendersParallel(templates);
+            return templates;
+        }
 
-            while (!downloadRenderTasks.All(x => x.IsCompleted) && !ctx.IsCancellationRequested)
-            {
-                await Task.Yield();
-            }
+        public async Task FetchTemplateRendersParallel(List<TemplateData> templates)
+        {
+            var tasks = templates.Select(FetchTemplateRenderAsync);
+            await Task.WhenAll(tasks);
```

Comment:
I am not sure if this is the right way, by opening so much api it confuses the usage of this class a bit. Though I understand why you did it. I have another proposition. So previously it was downloading all the templates json and downloading all the images parallelly but waiting for the download to complete and than returning. If you want to get the images as soon as they are downloaded, maybe you can add a callback(action) which is fired as soon as an image is downloaded.

Reply:
I see your point. I guess the main thing I am trying to do is not force devs to use it a particular way, enabling them to implement things however they like.

Reply:
But I guess they can already do that with API requests to get just template data only 
`return await avatarAPIRequests.GetTemplates();`

Reply:
I'm actually in favour of having the lower level layers (e.g. fetchers, or requests, or repositories as they'd be if we translate this to web) as open and flexible as possible, with the caveat that they stay within the bounds of their context entity so they don't get confusing (in this case 'templates'). The only thing I'd potentially change about this is a few small nitpicks that aren't crucial:

1. I wouldn't suffix the third method with Parallel. Since it's the only way of doing it we offer, it doesn't add any extra value to the user, and it sort of over explains the inner workings of the method. If we one day wanted to offer a way of linearly downloading all the template renders too (which would be a bit weird, but maybe 🤷‍♂️) then we use a strategy pattern and you'd pass in an enum of DownloadStrategy.Parallel, or DownloadStrategy.Linear. For now however, if it was me, I'd just had the method as `FetchTemplateRendersAsync`.

2. **_Maybe_** I'd split the template renders (bottom two methods) into a separate service / fetcher since the context is slightly different. However, I'm not as sure on this at the moment myself so just throwing it out there.

3. I thought we spoke about how GetTemplateAvatarImage wasn't really needed as all it really is in a image downloader? If you split this out into a generic image downloader then I think it'd become clearer how you might not even need this whole 'fetch template render' concept in here, since it's just fetching images...

4. I also like the prospect of a callback that's fired when a long running async thing is finished, but not sure how much value it would add here in this scenario 🤔